docs/design/adminconsole/INSTANCE-DATA.md

```diff
@@ -0,0 +1,280 @@
+  | odsInstanceDerivative.derivativeType | Max length: 50. Allowed values: "ReadReplica", "Snapshot". |
+
+* **Sample Payload**:
+
+  ```json
+  {
+    "odsInstanceId": 1,
+    "tenantId": 1,
+    "name": "Instance #1 - 2024",
+    "instanceType": "enterprise",
+    "odsInstanceContexts": [
+      {
+        "contextKey": "schoolYearFromRoute",
+        "contextValue": "2024"
+      }
+    ],
+    "odsInstanceDerivatives": [
+      {
+        "derivativeType": "ReadReplica"
+      }
+    ]
+  }
+  ```
+
+* **Response Format**:
+
+  ```json
+  {
+    "instanceId": 1,
+  }
+  ```
+
+### PUT /adminconsole/odsInstances/{id}
+
+* **Purpose**: Update an instance definition.
+* **Description**:
+  * Validate the incoming payload.
+  * Updates both the `adminconsole.Instance` and the `dbo.OdsInstances` tables.
+  * Respond with `202 No Content`
+* **Validation**:
+
+  | Property                             | Rules                                                      |
+  | ------------------------------------ | ---------------------------------------------------------- |
+  | name                                 | Max length: 100. Must be unique for the tenant.            |
+  | instanceType                         | Max length: 100.                                           |
+  | odsInstanceContexts                  | Empty array is allowed                                     |
+  | odsInstanceContext.contextKey        | Max length: 50. Must be unique for the instance.           |
+  | odsInstanceDerivatives               | Empty array is allowed                                     |
+  | odsInstanceContext.contextValue      | Max length: 50.                                            |
+  | odsInstanceDerivative.derivativeType | Max length: 50. Allowed values: "ReadReplica", "Snapshot". |
+
+* **Sample Payload**:
+
+  ```json
+  {
+    "odsInstanceId": 1,
+    "tenantId": 1,
+    "name": "Instance #1 - 2024",
+    "instanceType": "enterprise",
+    "odsInstanceContexts": [
+      {
+        "contextKey": "schoolYearFromRoute",
+        "contextValue": "2024"
+      }
+    ],
+    "odsInstanceDerivatives": [
+      {
+        "derivativeType": "ReadReplica"
+      }
+    ]
+  }
+  ```
+
+### DELETE /adminconsole/odsInstances/{id}
+
+* Not supported at this time. Respond with `405 Method Not Allowed`.
+
+### GET /adminconosle/instances
+
+Also supports `GET /adminconsole/instances/{id}`
+
+* **Purpose**: Provide instance list for the worker applications.
+* **Description**:
+  * Reads from the `adminconsole.Instance` table.
+  * Must be authorized with an appropriate Role name in the token.
+  * Returns a separate object for each ODS Instance Context.
+  * The `resourceUrl` is constructed from the tenant's base URL plus instance
+    context information.
+  * Respond with 200
+* **Query String**:
+  * `?status=` to search by Status.
+* **Response Format**:
+
+  ```json
+  [
+    {
```

Comment:
Should we add `BaseUrl` as well?

Reply:
In fact, I would prefer to add `baseUrl` rather than `tokenUrl` - should look up the Token URL using the base URL (also known as the Discovery API endpoint).

You don't need the `baseUrl` to perform the individual health checks; instead you would use the `resourceUrl`, since it has the additional routing and context information.

I am adding this as `discoveryApiUrl`.

Reply:
Done

Reply:
Just want to make sure we are on the same page. I don't think the discover `discoveryApiUrl` is enough. In a multi tenant environment and/or with ods context environment, the /instance payload doesn't have the data necessary to put the TokenUrl together. 
![discoverUrl](https://github.com/user-attachments/assets/deb60ce6-628a-4782-a470-4bfe6fdb38b8)
In this screenshot you can see what I am thinking the `ResourceUrl` and `TokenUrl` are.

Reply:
I'm still trying to figure out how to deploy multi-tenancy on my system to better understand how the Discovery API URL works. API clients are supposed to be able to find the token URL just by looking here, and it looks like we broke that promise.

In the meantime, I will update my document to reflect that we must provide the oauth URL.